src/common/utils/fold_test.cc

```diff
@@ -0,0 +1,96 @@
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "src/common/utils/fold.h"
+
+#include "absl/strings/str_join.h"
+#include "src/common/testing/gtest.h"
+
+namespace raksha::common::utils {
+
+struct IntVecAndSum {
+  std::vector<int> int_vec;
+  int sum;
+};
+
+class SumFoldTest : public testing::TestWithParam<IntVecAndSum> {};
+
+TEST_P(SumFoldTest, SumFoldTest) {
+  const auto &[vec, expected_result] = GetParam();
+  auto sum_fn = [](int sum_so_far, int num) { return sum_so_far + num; };
+  EXPECT_EQ(fold(vec, 0, sum_fn), expected_result);
+  EXPECT_EQ(rfold(vec, 0, sum_fn), expected_result);
+}
+
+static const IntVecAndSum kIntVecsAndSum[] = {
+    {.int_vec = {}, .sum = 0},
+    {.int_vec = {1}, .sum = 1},
+    {.int_vec = {1, 2, 3, 4}, .sum = 10},
+    {.int_vec = {-1, 5, 2, -8}, .sum = -2}};
+
+INSTANTIATE_TEST_SUITE_P(SumFoldTest, SumFoldTest,
+                         testing::ValuesIn(kIntVecsAndSum));
+
+class ConcatFoldTest : public testing::TestWithParam<std::vector<std::string>> {
+};
+
+TEST_P(ConcatFoldTest, ConcatFoldTest) {
+  const std::vector<std::string> &string_vec = GetParam();
+  auto concat_fn = [](std::string concat_so_far, const std::string &input) {
+    // Yes, this is bad form in general, but the quadraticness shouldn't matter
+    // for a test.
+    return concat_so_far + input;
+  };
+  EXPECT_EQ(fold(string_vec, std::string(""), concat_fn),
+            absl::StrJoin(string_vec, ""));
+  EXPECT_EQ(rfold(string_vec, std::string(""), concat_fn),
+            absl::StrJoin(string_vec.rbegin(), string_vec.rend(), ""));
+}
+
+static const std::vector<std::string> kSampleStringVecs[] = {
+    {}, {"x"}, {"hello", "world"}, {"a", "b", "c", "d", "e"}};
+
+INSTANTIATE_TEST_SUITE_P(ConcatFoldTest, ConcatFoldTest,
+                         testing::ValuesIn(kSampleStringVecs));
+
+// Box and unbox some strings via `fold`. This mainly tests that we can handle
+// an uncopyable accumulator (ie, the vector of unique pointers that we build
+// up) and an uncopyable input vector.
+class BoxFoldTest : public testing::TestWithParam<std::vector<std::string>> {};
+
+TEST_P(BoxFoldTest, BoxFoldTest) {
+  const std::vector<std::string> &string_vec = GetParam();
+  auto box_fn = [](std::vector<std::unique_ptr<std::string>> boxed_so_far,
+                   const std::string &input) {
+    boxed_so_far.push_back(std::make_unique<std::string>(input));
+    return boxed_so_far;
+  };
+  std::vector<std::unique_ptr<std::string>> boxed_strings =
+      fold(string_vec, std::vector<std::unique_ptr<std::string>>(), box_fn);
+  auto unbox_fn = [](std::vector<std::string> unboxed_so_far,
+                     const std::unique_ptr<std::string> &to_unbox) {
+    unboxed_so_far.push_back(*to_unbox);
+    return unboxed_so_far;
+  };
+  std::vector<std::string> unboxed_strings =
+      fold(boxed_strings, std::vector<std::string>(), unbox_fn);
```

Comment:
Doesn't passing `{}` for initial value work?
```suggestion
  std::vector<std::string> unboxed_strings =
      fold(boxed_strings, {}, unbox_fn);
```

Reply:
It's inferring the type provided in type parameter `A` from the passed-in `std::vector<std::string>` here. Using a `std::initializer_list` wouldn't provide type information and would make the compiler unhappy unless I explicitly provided type arguments.